### PR TITLE
Addresses #12456 

### PR DESCRIPTION
**Pull request by thehogfather**
Addresses issue where code folding triangles in the gutter were sometimes not correctly updated after deleting lines above a folded region.

### CODE REVIEW

**Comment by abose** [comment]
Tagging @swmitra

**Comment by ficristo** [comment]
I've looked a bit at this.
I was able to reproduce the issue (trying to delete a line of code above a folded region) and indeed this PR fixes it.
@thehogfather can you add a test?

I've noticed a small issue: if collapse a folded region and then expand it, the code folding triangles are not updated until you move the cursor.
fold

I think this is a minimal fix to solve the issue, but the right fix seems to me that we should update all the code of the code folding addon of CodeMirror.

**Comment by thehogfather** [comment]
@ficristo I will have a go at adding some more tests.

Could you clarify what you mean by 'update all the code of the code folding addon of CodeMirror'?

**Comment by ficristo** [comment]
If I'm not mistaken, the `foldcode.js` and `foldgutter.js` are a port of the CodeMirror ones adapted to work in Brackets. I think we should backport all the changes that are in CodeMirror (better in another PR).

**Comment by thehogfather** [comment]
Ah yes - better a different PR. There are lots of subtle differences, mainly to do with the ability to persist and restore fold states.

**Comment by thehogfather** [comment]
Changes ready for another review.
Main changes are updates to tests which now include html files  as well as js.

File: src/extensions/default/CodeFolding/foldhelpers/foldcode.js [hunk 1]
```diff
@@ -90,18 +90,24 @@ define(function (require, exports, module) {
...
-            CodeMirror.signal(cm, "unfold", cm, from, to, pos.line);
+            delete cm._lineFolds[from.line];
+            CodeMirror.signal(cm, "unfold", cm, from, to);
         });
 
         if (force === "fold") {
             delete range.cleared;
-            cm._lineFolds[pos.line] = range;
+            // in some cases such as in xml style files, the start of  line folds can span multiple lines
+            // render a gutter marker for both the beginning and end of the line
```
Comment (ficristo) [inline src/extensions/default/CodeFolding/foldhelpers/foldcode.js]: Use capital letter at the start of a sentence. This needs some dots too.

File: src/extensions/default/CodeFolding/unittest-files/test.html [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                     Choice 1
+                </label>
+                 <input type="radio" name="radio-choice-1" id="radio-choice-1" tabindex="2" value="choice-1" />
+
+                 <label for="radio-choice-2">Choice 2</label>
+                 <input type="radio" name="radio-choice-2" id="radio-choice-2" tabindex="3" value="choice-2" />
+            </div>
+        </form>
+    </body>
+</html>
```
Comment (ficristo) [inline src/extensions/default/CodeFolding/unittest-files/test.html]: Add a newline at the end of file please.
Are the empy lines necessary? If not remove.

File: src/extensions/default/CodeFolding/unittest-files/test.js [hunk 3]
```diff
@@ -1,33 +1,36 @@
...
-            return;
-        }
+  * Synchronises the code folding states in the CM doc to cm._lineFolds cache.
+  * When an undo operation is done, if folded code fragments are restored, then
+  * we need to update cm._lineFolds with the fragments
+  * @param {Object}   cm        cm the CodeMirror instance for the active  editor
+  * @param {Object}   from      starting position in the doc to sync the fold states from
+  * @param {[[Type]]} lineAdded a number to show how many lines where added to the document
+*/
+/*global define, brackets, document, window, $, CodeMirror, isFold, prefs*/
```
Comment (ficristo) [inline src/extensions/default/CodeFolding/unittest-files/test.js]: Put this as the first line followed by an empty line.
I think you can remove some of this too.
`isFold` and `prefs` cannot be required?

File: src/extensions/default/CodeFolding/unittest-files/test.js [hunk 4]
```diff
@@ -1,33 +1,36 @@
...
+
+function syncDocToFoldsCache(cm, from, lineAdded) {
+    "use strict";
+    var minFoldSize = prefs.getSetting("minFoldSize") || 2;
+    var opts = cm.state.foldGutter.options || {};
+    var rf = opts.rangeFinder || CodeMirror.fold.auto;
+    var i, pos, folds, fold, range;
```
Comment (ficristo) [inline src/extensions/default/CodeFolding/unittest-files/test.js]: Brackets style is to use a combined `var`.

File: src/extensions/default/CodeFolding/unittest-files/test.js [hunk 5]
```diff
@@ -1,33 +1,36 @@
...
+    if (lineAdded <= 0) {
+        return;
+    }
```
Comment (ficristo) [inline src/extensions/default/CodeFolding/unittest-files/test.js]: Can you put this directly aftrer `"use strict"`. I don't remember if JSLint complains.

File: src/extensions/default/CodeFolding/unittests.js [hunk 6]
```diff
@@ -22,8 +22,29 @@ define(function (require, exports, module) {
             foldMarkerOpen = gutterName + "-open",
             foldMarkerClosed = gutterName + "-folded";
         var extensionPath = FileUtils.getNativeModuleDirectoryPath(module),
-            testDocPath = extensionPath + "/unittest-files/",
-            testFilePath = testDocPath + "test.js";
+            testDocumentDirectory = extensionPath + "/unittest-files/",
+            testFiles = ["test.js", "test.hmtml"].map(function (file) {
```
Comment (ficristo) [inline src/extensions/default/CodeFolding/unittests.js]: I guess this should be `"test.html"`

File: src/extensions/default/CodeFolding/unittests.js [hunk 7]
```diff
@@ -22,8 +22,29 @@ define(function (require, exports, module) {
...
+            // The line numbers referenced below are dependent on the files in /unittest-files directory.
+            // Remember to update the numbers if the files change.
+            testFilesSpec = {
+                js: {
+                    filePath: testDocumentDirectory + "test.js",
+                    foldableLines: [ 11, 17, 21, 25, 27, 30],
+                    sameLevelFoldableLines: [17, 21],
+                    firstSelection: {start: {line: 2, ch: 0}, end: {line: 10, ch: 0}},
+                    secondSelection: {start: {line: 5, ch: 0}, end: {line: 8, ch: 4}}
+
```
Comment (ficristo) [inline src/extensions/default/CodeFolding/unittests.js]: Remove newline

**Comment by ficristo** [comment]
@thehogfather thank you a lot!
Could you merge / rebase with master? We have updated CodeMirror, so I would like to test with the new version.
Meanwhile I left some comments.

File: src/extensions/default/CodeFolding/unittest-files/test.html (hunk 2 above)
Comment (thehogfather) [inline src/extensions/default/CodeFolding/unittest-files/test.html]: This is a file used in the test as opposed to a regular source file. I would argue that this file should not need to conform to the brackets style  - but I am open to good reasons not to. For instance the line break on line 27 is intentional - it is a special case of when the fold gutter appears on a line different from when the beginning of the range starts.

File: src/extensions/default/CodeFolding/unittest-files/test.js (hunk 3 above)
Comment (thehogfather) [inline src/extensions/default/CodeFolding/unittest-files/test.js]: Similar to the html, this is a test file loaded up during the unit tests. the `isFold` and `prefs` are just there to silence jshint errors.

File: src/extensions/default/CodeFolding/unittest-files/test.js (hunk 4 above)
Comment (thehogfather) [inline src/extensions/default/CodeFolding/unittest-files/test.js]: Similarly, I dont think this file should conform to the brackets style - they are test files and should be able to accommodate plausible scenarios that might appear in projects of anyone using brackets regardless of style.

File: src/extensions/default/CodeFolding/unittests.js (hunk 6 above)
Comment (thehogfather) [inline src/extensions/default/CodeFolding/unittests.js]: good catch. This must be a redundant line. Will remove.

**Comment by ficristo** [comment]
If the newlines I mentioned are necessary as the globals to make testing easier, then is fine to leave them.

**Comment by ficristo** [comment]
@thehogfather could you take a look at the test failures?

So now when a tag span on multiple lines you have two code folding triangles. I'm not an UI / UX expert so I'm not sure is actually a good or bad idea.
html-foldable

/cc @MarcelGerber

**Comment by thehogfather** [comment]
@ficristo regarding the folding triangles, I think this is fine. A better design might be to hide the opening tag when the range is collapsed or to only show the fold triangles at the start of the folded/foldable range (so in your example we dont show line 25 as foldable - only line 29). This is however a little tricky - would need a rewrite of the xml-range finder. We are currently using whatever version is from code-mirror. The same workaround is present in code-mirror. I think it is ok as it doesn't reduce functionality nor does it put the editor in an unusable state.

**Comment by redmunds** [comment]
Another option would be to collapse both tag attributes and tag contents something like this:

tag-collapse2

**Comment by ficristo** [comment]
I still see some test failures on the latest commit.

**Comment by thehogfather** [comment]
@ficristo would you kindly point out the failing test if possible? I typically run the extension test as opposed to 'All tests' so I might have missed something. I've just run the tests again and all are passing. Note that the last commits need a full reload of brackets so that the changes in code mirror are reflected in the tests.

**Comment by thehogfather** [comment]
@redmunds interesting idea - I'd however question  what happens to the fold mark in the gutter when either of the collapsed ranges in the editor is clicked. They should probably be treated as two independent ranges (as displayed in the editor) and marked as such in the gutter with two fold triangles when open.

**Comment by ficristo** [comment]
Since the behaviour match the CodeMirror one I'm going to take this as is and I'll open a new issue to discuss more the UI / UX.
I've run multiple times the tests and seems there are some issues with the focus of the test window: when I give the focus on it all the tests pass.

@thehogfather can you squash a bit the commits?

**Comment by thehogfather** [comment]
@ficristo not sure how much more I can squash this. I have tried without much success. Maybe whoever merges on github can use the Squash and merge button? Alternatively I would probably create a new PR with all the changes if that is preferable.

**Comment by ficristo** [comment]
@thehogfather thank you.
I've merged as is (I didn't want to mess with github)

**Pull request merged**